communitiesuk/funding-service-design-post-award-data-store
# FMD-111: Database restructure

https://dluhcdigital.atlassian.net/browse/FMD-111
https://dluhcdigital.atlassian.net/browse/FMD-112
https://dluhcdigital.atlassian.net/browse/FMD-113
https://dluhcdigital.atlassian.net/browse/FMD-114


### Change description

Updates DB model definition file to include ProgrammeJunction.
Add mappings to re-route all Programme level data through this junction table.
Updates all relevant Indexes, ORM relationships, Mappings etc
Removes old FK structures including obsolete FK's to submission from Event data tables

DB / Data migration file to perform the above changes on the live database

Updates Ingest pipeline/methods to insert data into the new structure
Historical ingest pipelines updated to work with this structure.

Updates Serialisation / file download methods and queries to work with the new structure

Extensive test updates to ensure existing test logic still all works with the new structure. Including:
- Fixture changes to seed process, including source data csv mapping changes to include updated FK relations
- Serialisation tests
- Integration tests
- Extraction tests
- historical ingest/extraction tests
- DB tests (constraints, hybrid_properties, queries)




### How to test
- To test app normally, pull changes, and run tests.
- To manually test (run locally), pull changes, make sure DB migrations are applied, run app. This may require a DB reset from docker-runner scripts if you have an existing DB with lots of test data in it.
- To test the data migration locally, a useful process is:
    - Switch to main branch
    - Run Flask db upgrade to previous alembic migration.
    - Add data to the DB by either/both of seed CLI or ingest endpoint
    - Download extract spreadsheet for comparison, (if required)
    - Switch back to feature branch
    - Run flask db upgrade 
    - Make sure this runs without error
    - Run app, download extract to compare (or whatever checks you want to do). The data should have migrated and for all intents and purposed appear the same as before any changes were made.
    - Optionally you could run the downgrade of the migration, and re-checkout main, and ensure it still all works the same

The above long testing process will need to be done in some detail (for QA) on a realistic test env before deploying any of these changes to prod.

### ERD (abstracted) for general DB structure after migration.

![Find_ERD_proposed_V2 1](https://github.com/communitiesuk/funding-service-design-post-award-data-store/assets/95703578/d7f4403c-5532-487c-94c5-a401c15fb42d)
